src/services/rollbackService.ts

```diff
@@ -0,0 +1,146 @@
+  private blobService: AzureBlobStorageService;
+
+  /**
+   * Initialize rollback service, including authentication and initialization
+   * of a `ResourceService`
+   * @param serverless Serverless object
+   * @param options Serverless CLI options
+   */
+  public constructor(serverless: Serverless, options: Serverless.Options) {
+    super(serverless, options);
+    this.resourceService = new ResourceService(serverless, options);
+    this.blobService = new AzureBlobStorageService(serverless, options);
+  }
+
+  /**
+   * Rolls back the function app. If `timestamp` present in `options`,
+   * function app will be rolled back directly to that timestamp. Otherwise,
+   * function app will be rolled back to the deployment previous to the most recent
+   */
+  public async rollback() {
+    // Get deployment as specified by timestamp
+    const deployment = await this.getDeployment();
+    if (!deployment) {
+      return;
+    }
+    // Name of artifact in blob storage
+    const artifactName = this.getArtifactName(deployment.name);
+    // Redeploy resource group (includes SAS token URL if running from blob URL)
+    await this.redeployDeployment(deployment, artifactName);
+  }
+
+  /**
+   * Re-deploy a previous deployment of a resource group
+   * @param deployment Previous deployment retrieved from Azure SDK
+   * @param artifactName Name of zipped artifact in blob storage associated with deployment
+   */
+  private async redeployDeployment(deployment: DeploymentExtended, artifactName: string) {
+    const armService = new ArmService(this.serverless);
+    const armDeployment = await this.convertToArmDeployment(deployment);
+    // Initialize blob service for either creating SAS token or downloading artifact to uplod to function app
+    await this.blobService.initialize();
+    if (this.deploymentConfig.runFromBlobUrl) {
+      // Set functionRunFromPackage param to SAS URL of blob
+      armDeployment.parameters.functionAppRunFromPackage = await this.blobService.generateBlobSasTokenUrl(
+        this.deploymentConfig.container,
+        artifactName
+      )
+    }
+    await armService.deployTemplate(armDeployment);
+    /**
+     * Cannot use an `else` statement just because deploying the artifact
+     * depends on `deployTemplate` already being called
+     */
+    if (!this.deploymentConfig.runFromBlobUrl) {
+      const artifactPath = await this.downloadArtifact(artifactName);
+      await this.redeployArtifact(artifactPath);
+    }
+  }
+
+  /**
+   * Convert previous deployment to ArmDeployment
+   * @param deployment Previous deployment retrieved from Azure SDK
+   */
+  private async convertToArmDeployment(deployment: DeploymentExtended): Promise<ArmDeployment> {
+    const resourceService = new ResourceService(this.serverless, this.options);
+    const { template } = await resourceService.getDeploymentTemplate(deployment.name);
+    return {
+      template,
+      parameters: deployment.properties.parameters,
```

Comment:
If you convert the params back to the expected format here you can remove the special case in the `ArmService` that you had asked me about.

Reply:
will do. thanks